functions/go/gatekeeper/README.md

```diff
@@ -4,29 +4,87 @@
 
 <!--mdtogo:Short-->
 
-Validate the KRM resources using [Gatekeeper] policies.
+Validate the KRM resources using [Gatekeeper] policies. The `gatekeeper`
+function wraps the `Gatekeeper` policy enforcement engine.
+
+[Gatekeeper] provides an extensible policy enforcement engine which allows users
+to validate the KRM resources against the Gatekeeper policies.
+
+There are 2 kinds of resources needed to define a policy:
+- [Constraint Template]: Define the schema and logic of a policy. The policy
+  logic in a Constraint Template must be written in the [Rego] language.
+- [Constraint]: Signal the Gatekeeper the corresponding constraints need
+  to be enforced. Every Constraint must be backed by a Constraint Template.
+
+To make it do something meaningful, you should to define at least one
+[Constraint Template] and at least one [Constraint]. If you don't provide any
+[Constraint Template] or [Constraint], it's a NOOP for the function.
+You can learn more about how to use the `Gatekeeper` project [here][howto]. The
+concept document about the `OPA Constraint Framework` can be found [here][concept].
 
 <!--mdtogo-->
 
+<!--mdtogo:Long-->
+
 ### FunctionConfig
 
-<!--mdtogo:Long-->
+The constraint templates and the constraints must be provided using
+`input items` along with other KRM resources to be validated. Nothing need to be
+provided in `input functionConfig`.
```

Comment:
Not sure what `input functionConfig` is.

Showing the CRD in the section called `FunctionConfig` is weird as they's no FunctionConfig. 

The biggest issue is that there are holes in story here, what the reader is looking for from this doc:

- What's this function about?
- What problem does it solve?
- Why should I use it?
- How should I use it?

This PR doesn't actually expand on any of this.

Reply:
I am looking at the function-catalog docs (especially functions that wraps other tools) and realizing that there are two types of audience for such docs:

- Users who don't know about the tools being wrapped. This audience will actually start with a higher level problem such as "I want to ensure some form of extra validation (constraints, policies....) for my configs", so they need something like I would call a `solution page`.
- Users who know what the wrapped tool does. They also know how `kpt` functions work etc. So all they need is a quick reference docs and some examples. Solution page will also refer to these reference page.

So question I am trying to answer in my head is: what does this page mean ? is it `reference page` or `solution page`. It seems it is trying to do both the things and that can get confusing for the author of the docs as well as audience of the docs.

Reply:
This is a higher level discussion then.
I think we can iterate on the docs.
The revision in this PR is better than what we currently have. I guess we can merge this PR soon-ish and then figure out if we want to have reference page and solution page. And how.

Reply:
Agree. I think we can have a separate `tutorial.md` for each function and have two links on the function homepage (shown below). I am treating this page more as a reference page for now.

[tutorial](..)|[reference](...)